File: benchmarks/camada-de-execucao/io_client.mjs
// Micro-problema 2 — lado Node. Mesmo trabalho do io_client.go: K jobs em voo,
// cada um assinado e enviado ao provedor falso. Node não tem goroutine; a
// concorrência vem do laço de eventos, com um pool de K promessas.
import fs from "node:fs";
import http from "node:http";
import { createHash, createHmac } from "node:crypto";

const total = Number(process.argv[2]);
const concorrencia = Number(process.argv[3]);
const corpo = fs.readFileSync(new URL("./payload.json", import.meta.url));
const SEGREDO = "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef";

// Equivalente ao http.Transport do Go: reuso de conexão com teto igual.
const agente = new http.Agent({ keepAlive: true, maxSockets: concorrencia, maxFreeSockets: concorrencia });

function rssPicoKB() {
  for (const linha of fs.readFileSync("/proc/self/status", "utf8").split("\n")) {
    if (linha.startsWith("VmHWM:")) return Number(linha.split(/\s+/)[1]);
  }
  return 0;
}

function enviarUm(i) {
  const canonica = [
    String(Date.now()),
    "nonce-" + i,
    "POST",
    "/v1/messages/dispatch",
    createHash("sha256").update(corpo).digest("hex")
  ].join(".");
  const assinatura = createHmac("sha256", SEGREDO).update(canonica).digest("hex");

  return new Promise(resolve => {
    const req = http.request({
      host: "127.0.0.1", port: 8791, path: "/v1/messages/dispatch", method: "POST",
      agent: agente,
      headers: {
        "Content-Type": "application/json",
        "Content-Length": corpo.length,
        "x-innov-signature": assinatura
      }
    }, resp => {
      resp.resume();
      resp.on("end", () => resolve(true));
    });
    req.on("error", () => resolve(false));
    req.end(corpo);
  });
}

let proximo = 0, ok = 0, falhas = 0;
const inicio = process.hrtime.bigint();

await Promise.all(Array.from({ length: concorrencia }, async () => {
  while (proximo < total) {
    const i = proximo++;
    if (await enviarUm(i)) ok++; else falhas++;
  }
}));

const segundos = Number(process.hrtime.bigint() - inicio) / 1e9;
console.log(JSON.stringify({
  runtime: "node " + process.versions.node,
  total, concorrencia,
  segundos: segundos.toFixed(3),
  jobs_por_seg: Math.round(ok / segundos),
  ok, falhas,
  rss_pico_kb: rssPicoKB()
}));

File: benchmarks/camada-de-execucao/io_client.go
// Micro-problema 2 — quantos jobs em voo o runtime sustenta, e a que custo de
// memória. É o que a camada de execução faz de verdade: manter K chamadas ao
// provedor simultâneas, assinando cada uma. Go: uma goroutine por job.
package main

import (
	"bytes"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"runtime"
	"strconv"
	"strings"
	"sync"
	"sync/atomic"
	"time"
)

const segredo = "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef"

func rssKB() int64 {
	b, err := os.ReadFile("/proc/self/status")
	if err != nil {
		return 0
	}
	for _, linha := range strings.Split(string(b), "\n") {
		if strings.HasPrefix(linha, "VmHWM:") {
			campos := strings.Fields(linha)
			v, _ := strconv.ParseInt(campos[1], 10, 64)
			return v
		}
	}
	return 0
}

func main() {
	total, _ := strconv.Atoi(os.Args[1])
	concorrencia, _ := strconv.Atoi(os.Args[2])
	corpo, _ := os.ReadFile("payload.json")

	transporte := &http.Transport{
		MaxIdleConns:        concorrencia,
		MaxIdleConnsPerHost: concorrencia,
		MaxConnsPerHost:     concorrencia,
		IdleConnTimeout:     90 * time.Second,
	}
	cliente := &http.Client{Transport: transporte, Timeout: 30 * time.Second}

	var ok, falhas int64
	jobs := make(chan int, total)
	for i := 0; i < total; i++ {
		jobs <- i
	}
	close(jobs)

	inicio := time.Now()
	var wg sync.WaitGroup
	for w := 0; w < concorrencia; w++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for i := range jobs {
				soma := sha256.Sum256(corpo)
				canonica := strings.Join([]string{
					strconv.FormatInt(time.Now().UnixMilli(), 10),
					"nonce-" + strconv.Itoa(i),
					"POST", "/v1/messages/dispatch", hex.EncodeToString(soma[:]),
				}, ".")
				mac := hmac.New(sha256.New, []byte(segredo))
				mac.Write([]byte(canonica))
				assinatura := hex.EncodeToString(mac.Sum(nil))

				req, _ := http.NewRequest("POST", "http://127.0.0.1:8791/v1/messages/dispatch", bytes.NewReader(corpo))
				req.Header.Set("Content-Type", "application/json")
				req.Header.Set("x-innov-signature", assinatura)
				resp, err := cliente.Do(req)
				if err != nil {
					atomic.AddInt64(&falhas, 1)
					continue
				}
				io.Copy(io.Discard, resp.Body)
				resp.Body.Close()
				atomic.AddInt64(&ok, 1)
			}
		}()
	}
	wg.Wait()
	decorrido := time.Since(inicio)

	saida, _ := json.Marshal(map[string]any{
		"runtime": "go " + runtime.Version(), "total": total, "concorrencia": concorrencia,
		"segundos":     fmt.Sprintf("%.3f", decorrido.Seconds()),
		"jobs_por_seg": int64(float64(ok) / decorrido.Seconds()),
		"ok":           ok, "falhas": falhas, "rss_pico_kb": rssKB(),
	})
	fmt.Println(string(saida))
}
